ios/safe-area/RNSSafeAreaViewComponentView.mm

```diff
@@ -23,44 +48,162 @@ - (instancetype)initWithFrame:(CGRect)frame
   }
   return self;
 }
+#else
+- (instancetype)initWithBridge:(RCTBridge *)bridge
+{
+  if (self = [super initWithFrame:CGRectZero]) {
+    _bridge = bridge;
+    [self initState];
+  }
+
+  return self;
+}
+
+RCT_NOT_IMPLEMENTED(-(instancetype)initWithCoder : (NSCoder *)decoder)
+RCT_NOT_IMPLEMENTED(-(instancetype)initWithFrame : (CGRect)frame)
+#endif // RCT_NEW_ARCH_ENABLED
 
 - (void)initState
 {
 #if RCT_NEW_ARCH_ENABLED
   static const auto defaultProps = std::make_shared<const react::RNSSafeAreaViewProps>();
   _props = defaultProps;
+#else
+  _edges = RNSSafeAreaViewEdgesMake(false, false, false, false);
 #endif // RCT_NEW_ARCH_ENABLED
+}
+
+- (void)didMoveToWindow
+{
+  UIView *previousProviderView = _providerView;
+  _providerView = [self findNearestProvider];
 
-  //  [self resetProps];
+  [self invalidateSafeAreaInsets];
+
+  if (previousProviderView != _providerView) {
+    if (previousProviderView != nil) {
+      [NSNotificationCenter.defaultCenter removeObserver:self name:RNSSafeAreaDidChange object:previousProviderView];
+    }
+    if (_providerView != nil) {
+      [NSNotificationCenter.defaultCenter addObserver:self
+                                             selector:@selector(safeAreaProviderInsetsDidChange:)
+                                                 name:RNSSafeAreaDidChange
+                                               object:_providerView];
+    }
+  }
 }
 
-//- (void)resetProps
-//{
-//  return;
-//}
+- (UIView<RNSSafeAreaProviding> *_Nullable)findNearestProvider
+{
+  UIView *current = self.superview;
+  while (current != nil) {
+    if ([current respondsToSelector:@selector(RNS_safeAreaInsets)]) {
+      return static_cast<UIView<RNSSafeAreaProviding> *>(current);
+    }
+    current = current.superview;
+  }
+  return nil;
+}
+
+- (void)safeAreaProviderInsetsDidChange:(NSNotification *)notification
+{
+  [self invalidateSafeAreaInsets];
+}
+
+- (void)invalidateSafeAreaInsets
+{
+  if (_providerView == nil) {
+    return;
+  }
+
+  UIEdgeInsets safeAreaInsets = _providerView.RNS_safeAreaInsets;
 
+  if (UIEdgeInsetsEqualToEdgeInsetsWithThreshold(safeAreaInsets, _currentSafeAreaInsets, 1.0 / RCTScreenScale())) {
+    return;
+  }
+
+  _currentSafeAreaInsets = safeAreaInsets;
 #if RCT_NEW_ARCH_ENABLED
-#pragma mark - RCTViewComponentViewProtocol
+  [self updateState];
+#else
+  [self updateLocalData];
+#endif // RCT_NEW_ARCH_ENABLED
+}
+
+#if RCT_NEW_ARCH_ENABLED
+- (void)updateState
+{
+  using facebook::react::RNSSafeAreaViewShadowNode;
+
+  if (!_state) {
+    return;
+  }
+
+  _state->updateState(
+      [=](RNSSafeAreaViewShadowNode::ConcreteState::Data const &oldData)
+          -> RNSSafeAreaViewShadowNode::ConcreteState::SharedData {
+        auto newData = oldData;
+        newData.insets = RCTEdgeInsetsFromUIEdgeInsets(_currentSafeAreaInsets);
+        return std::make_shared<RNSSafeAreaViewShadowNode::ConcreteState::Data const>(newData);
+      });
```

Comment:
Why do we use this more verbose form of updateState here, instead of just passing the payload?